grant/email: addded change owner template

thelostone-mc opened the pull request:
##### Description

Demo: https://gitcoincore.slack.com/files/UAW6336F2/FEYNYNGR3/_derp.mov

```
- `A` -> changes grant ownership to `B` 
- we keep record of that and send a mail to `B` 
- `B` can choose to accept or reject
- If `B` accepts -> he is redirected to grant page with alert say he has rejected the request & a mail gets sent to him saying the same
- If `B` rejects -> he is redirected to grant page with alert notifying he is now the owner & a mail gets sent to `A` saying the change was rejected
```

For @mbeacom / @captnseagraves 

I added a new field in model `request_ownership_change` which gets populates when owners tries to change owner. Once the potentail new owner accepts / rejects  -> we update the `admin_profile` and set `request_ownership_change` as None

- [x] verify if the email calls are done right 
    [request new owner](https://github.com/gitcoinco/web/pull/3252/files#diff-9c83a1892e5b433bc22483a941276b94R145) 
   [new owner rejects](https://github.com/gitcoinco/web/pull/3252/files#diff-9c83a1892e5b433bc22483a941276b94R184)
   [new owner accepts](https://github.com/gitcoinco/web/pull/3252/files#diff-9c83a1892e5b433bc22483a941276b94R179)
- [x] pass this [param](https://github.com/gitcoinco/web/pull/3252/files#diff-9c83a1892e5b433bc22483a941276b94R146) to the redirect so that [this alert](https://github.com/gitcoinco/web/pull/3252/files#diff-405c08ff03effe01b34326d2175ff1beR374) is rendered notifying the current owner that an email has been sent out 
- [x] should this [change ownership approval mail](https://github.com/gitcoinco/web/pull/3252/files#diff-9c83a1892e5b433bc22483a941276b94R178) be sent to the previous owner ?

thelostone-mc commented:
@owocki / @captnseagraves could use a hand with this !
Else we merge and figure it out later ?

owocki commented:
well need to rebase on the latest master so the builds succeed

owocki commented on app/grants/migrations/0012_grant_request_ownership_change.py:
```diff
@@ -0,0 +1,20 @@
+# Generated by Django 2.1.2 on 2018-12-19 21:18
+
+from django.db import migrations, models
+import django.db.models.deletion
```
i think the migrations have been squashed since this was written!

owocki commented on app/grants/views.py:
```diff
@@ -130,12 +130,15 @@ def grant_details(request, grant_id, grant_slug):
+            if grant.admin_profile != admin_profile:
+                grant.request_ownership_change = admin_profile
+                change_grant_owner(grant, grant.request_ownership_change)
+                # TODO : Pass { 'change_ownership' = 'R' } to template
```
why is this a TODO?  will need to be resolved before deploy?

owocki commented on app/grants/views.py:
```diff
@@ -160,6 +163,22 @@ def grant_details(request, grant_id, grant_slug):
+            grant.admin_profile = grant.request_ownership_change
+            grant.request_ownership_change = None
+            grant.save()
+            # Should mail be sent to the previous owner also ?
```
yes i think so

owocki commented on app/marketing/mails.py:
```diff
@@ -122,6 +123,51 @@ def new_grant(grant, profile):
         translation.activate(cur_language)
 
 
+def change_grant_owner(grant, profile):
```
`change_grant_owner_request` might be a better naming convention?

owocki commented on app/retail/templates/emails/grants/change_owner.html:
```diff
@@ -0,0 +1,202 @@
+{% block content %}
+
+<style>
+  h1 {
```
it'd be good to DRY this CSS if possible.

owocki commented on app/retail/templates/emails/grants/change_owner.html:
```diff
@@ -0,0 +1,202 @@
+</p>
+<hr>
+
+<div class="grant-info">
```
can we DRY up the common modules at all?

thelostone-mc commented:
@owocki 

- Added email to notify old owner when ownership changes
- renamed the method as requested 
- I removed the migration file and created it again. seems to be having  some weird issues
If I squash all the migration in grants -> it works fine. 
Can I push that commit in / would that mess things up? 
<img width="638" alt="screen shot 2019-01-11 at 7 23 55 pm" src="https://user-images.githubusercontent.com/5358146/51037576-9ba55880-15d6-11e9-964e-b4033da6f9c2.png">

will do this in a followup PR : 
- DRY up the css across grant mails.

what I need help with is  (chec issue description )
> pass this param to the redirect so that this alert is rendered notifying the current owner that an email has been sent out

aka https://github.com/gitcoinco/web/pull/3366#discussion_r246601211

owocki commented:
@thelostone-mc i think if you squash the migrations, that will screw things up in prod because the migrations are not squashed there.

I'm not sure how Mark did the migration squshing in prod before..

what is the 'weird issue' you're seeing?

thelostone-mc commented:
@owocki that's what I figured 
If you pull this in fresh and run migrate -> It woold throw this error

```
web_1      |     return self.cursor.execute(sql, params)
web_1      | django.db.utils.ProgrammingError: column grants_grant.request_ownership_change_id does not exist
web_1      | LINE 1: ...ed_gas_price", "grants_grant"."admin_profile_id", "grants_gr...
```

owocki commented:
@thelostone-mc if you run `./manage.py makemigrations` it should make a migration that solves that error... (unless your DB state is corrupted by working on this PR and moving around the migrations)..

owocki commented:
As an aside, I find it very confusing to squash the migrations.  Having 100 migrations doesnt cost us anything.  Having to wrestle with incosistent migration history does.

thelostone-mc commented:
I'll give it another shot! But yeah I did run make migrations to recreate the file. 
Oh it just slows down there amount of time or app takes to build up when you start fresh :P that's why we did it in the first place

thelostone-mc commented:
Steps to test: 

```
- navigate to grant & change owner
- you will see a notification saying your request has been sent 
-  login as the other user and visit http://localhost:8000/_administration/email/change_grant_owner_request to accept or deny the request  (if you have issues simply add `request_ownership_change` to admin.py and update

- If you accept 
user is directed to grant -> alert saying you are not the grant owner
new owner email : http://localhost:8000/_administration/email/change_grant_owner_accept
previous owner email : http://localhost:8000/_administration/email/change_grant_owner_accept


- If you reject
user is directed to grant -> alert saying you have denied the request for ownership transfer
current owner email : http://localhost:8000/_administration/email/change_grant_owner_reject
```

owocki commented on app/grants/views.py:
```diff
@@ -160,6 +163,22 @@ def grant_details(request, grant_id, grant_slug):
     }
+
+    if request.method == 'GET' and grant.request_ownership_change and profile == grant.request_ownership_change:
+        # if request.method == 'GET' and grant.request_ownership_change:
```
dead code

SaptakS commented on app/grants/templates/grants/detail.html:
```diff
@@ -380,4 +380,16 @@ <h5 class="font-body mt-2 font-weight-semibold">
+      _alert({ message: gettext('Ownership Request has been rejected.') }, 'error');
+    {% endif %}
+
+    if(localStorage['request_change'] == 'R') {
```
Space after if. And why not `===` ?

SaptakS commented on app/grants/views.py:
```diff
@@ -160,6 +163,22 @@ def grant_details(request, grant_id, grant_slug):
+
+    if request.method == 'GET' and grant.request_ownership_change and profile == grant.request_ownership_change:
+        # if request.method == 'GET' and grant.request_ownership_change:
+        if request.GET.get('ownership') == 'accept':
```
preferably `request.GET.get('ownership', None)`

SaptakS commented on app/marketing/mails.py:
```diff
@@ -122,6 +124,57 @@ def new_grant(grant, profile):
+        html, text, subject = render_change_grant_owner_request(grant)
+
+        if not should_suppress_notification_email(to_email, 'change_owner'):
+            send_mail(from_email, to_email, subject, text, html, categories=['transactional', func_name()])
```
Why not `except` ? You should catch the error and log in the error logger. So we can track what is wrong if need be.

SaptakS commented on app/retail/emails.py:
```diff
@@ -77,6 +77,38 @@ def render_new_grant_email(grant):
+    params = {'grant': grant}
+    response_html = premailer_transform(render_to_string("emails/grants/change_owner_request.html", params))
+    response_txt = render_to_string("emails/grants/change_owner_request.txt", params)
+    subject = "You've been chosen to be the owner for a Gitcoin Grant"
```
`subject = _("You've been chosen to be the owner for a Gitcoin Grant")` ?

SaptakS commented on app/retail/emails.py:
```diff
@@ -128,7 +160,7 @@ def render_successful_contribution_email(grant, subscription, contribution):
 @staff_member_required
 def successful_contribution(request):
-    grant = Grant.objects.all().first()
+    grant = Grant.objects.first()
```
🙌

thelostone-mc commented:
@SaptakS resolved feedback [here](https://github.com/gitcoinco/web/pull/3366/commits/5d6c217eac0d17c9b4c079707fbf41ff39cffb85)
